src/markov/markov_approx.jl

```diff
@@ -219,3 +223,567 @@ function estimate_mc_discrete{T}(X::Vector{T})
+
+    Tx = T(D)
+    L, N2 = size(Tx)
+
+    if N2 != N || length(TBar) != L
+        error("Dimension mismatch")
+    end
+
+    # Compute maximum entropy discrete distribution
+    options = Optim.Options(f_tol=1e-16, x_tol=1e-16)
+    obj(lambda) = entropy_obj(lambda, Tx, TBar, q)
+    grad!(lambda, grad) = entropy_grad!(grad, lambda, Tx, TBar, q)
+    hess!(lambda, hess) = entropy_hess!(hess, lambda, Tx, TBar, q)
+    res = Optim.optimize(obj, grad!, hess!, lambda0, Optim.Newton(), options)
+    # Sometimes the algorithm fails to converge if the initial guess is too far
+    # away from the truth. If this occurs, the program tries an initial guess
+    # of all zeros.
+    if !Optim.converged(res) && all(lambda0 .!= 0.0)
+        warn("Failed to find a solution from provided initial guess. Trying new initial guess.")
+        res = Optim.optimize(obj, grad!, hess!, zeros(lambda0), Optim.Newton(), options)
+        # check convergence
+        Optim.converged(res) || error("Failed to find a solution.")
+    end
+    # Compute final probability weights and moment errors
+    lambdaBar = Optim.minimizer(res)
+    minimum_value = Optim.minimum(res)
+
+    Tdiff = Tx-repmat(TBar, 1, N)
+    p = (q.*exp.(lambdaBar'*Tdiff))/minimum_value
+    grad = similar(lambda0)
+    grad!(Optim.minimizer(res), grad)
+    momentError = grad/minimum_value
+    return p, lambdaBar, momentError
+end
+
+"""
+
+Compute the moment defining function used in discrete_approximation
+
+```julia
+T = polynomial_moment(X, mu, scalingFactor, mMoments)
+```
+
+##### Argumentss:
+
+- `X::Vector` : (N x 1) vector of grid points
+- `mu::Real` : location parameter (conditional mean)
+- `scalingFactor::Real` : scaling factor for numerical stability
+                          (typically largest grid point)
+- `nMoments::Integer` : number of polynomial moments
+
+##### Return
+
+- `T` : moment defining function used in discrete_approximation
+
+"""
+function polynomial_moment(X::Vector, mu::Real, scalingFactor::Real, nMoments::Integer)
+    # Check that scaling factor is positive
+    scalingFactor>0 || error("scalingFactor must be a positive number")
+    Y = (X-mu)/scalingFactor # standardized grid
+    T = Y'.^collect(1:nMoments)
+end
+
+"""
+
+Compute the maximum entropy objective function used in discrete_approximation
+
+```julia
+obj = entropy_obj(lambda, Tx, TBar, q)
+```
+
+##### Arguments
+
+- `lambda::Vector` : (L x 1) vector of values of the dual problem variables
+- `Tx::Matrix` : (L x N) matrix of moments evaluated at the grid points
+         specified in discrete_approximation
+- `TBar::Vector` : (L x 1) vector of moments of the underlying distribution
+           which should be matched
+- `q::Matrix` : (1 X N) vector of prior weights for each point in the grid.
+
+##### Returns
+
+- `obj` : scalar value of objective function evaluated at lambda
+
+"""
+function entropy_obj(lambda::Vector, Tx::Matrix, TBar::Vector, q::Matrix)
+
+    # Some error checking
```

Comment:
Is `entropy_obj` something users will call by themselves, or just a internal routine we use in `discrete_approximation`?

If it is meant to be internal, could we do the error checking just once from within `discrete_approximation` so ti doesn't happen every iteration of the solver?